ReSharper.FSharp/src/FSharp/FSharp.Psi.Services/src/Util/TypeAnnotationsUtil.fs

```diff
@@ -0,0 +1,90 @@
+module JetBrains.ReSharper.Plugins.FSharp.Psi.Services.Util.TypeAnnotationsUtil
+
+open JetBrains.ReSharper.Plugins.FSharp.Psi.Tree
+open JetBrains.ReSharper.Psi.Tree
+
+let rec private visitPattern (acc: ITreeNode list) (pattern: IFSharpPattern) =
+    match pattern with
+    | null
+    | :? IUnitPat
+    | :? IWildPat
+    | :? ITypedPat -> acc
+
+    | :? IParenPat as parenPat -> visitPattern acc parenPat.Pattern
+    | :? IAsPat as asPat -> visitPattern acc asPat.LeftPattern
+
+    | :? IReferencePat as pat -> pat :: acc
+
+    | :? IRecordPat as recordPat ->
+        recordPat.FieldPatternsEnumerable
+        |> Seq.fold visitPattern acc
+
+    | :? IFieldPat as fieldPat -> visitPattern acc fieldPat.Pattern
+    | :? IAttribPat as attribPat -> visitPattern acc attribPat.Pattern
+
+    | :? ITuplePat as tuplePat ->
+        tuplePat.PatternsEnumerable
+        |> Seq.fold visitPattern acc
+
+    | :? IParametersOwnerPat as parametersOwnerPat ->
+        let acc = (parametersOwnerPat : ITreeNode) :: acc
+        parametersOwnerPat.ParametersEnumerable
+        |> Seq.fold visitPattern acc
+
+    | :? INamedUnionCaseFieldsPat as unionCaseFieldsPat ->
+        unionCaseFieldsPat.FieldPatternsEnumerable
+        |> Seq.fold visitPattern acc
+
+    | _ -> acc
+
+let private collectPatternsRequiringAnnotations acc (parametersOwner: IParameterOwnerMemberDeclaration) =
+    parametersOwner.ParameterPatterns
+    |> Seq.fold visitPattern acc
+
+let collectTypeHintAnchorsForBinding (binding: IBinding) =
+    let acc: ITreeNode list = []
+
+    if isNull binding || isNull binding.EqualsToken then acc else
+
+    let acc =
+        if isNotNull binding.ReturnTypeInfo then acc
+        elif binding.HasParameters then [(binding: ITreeNode)]
```

Comment:
Are parens needed here?

Reply:
Yeah 

![{08508C3B-125C-4CC3-B53C-A395ADD78022}](https://github.com/user-attachments/assets/95869d03-0763-4098-9972-81bdb9f63948)

![{651D424A-581C-4B9F-8593-01C0C2B2737A}](https://github.com/user-attachments/assets/9b272788-b3c7-4e77-8b6e-8e114f18efab)


it is also possible to write `binding :> ITreeNode`

Reply:
Wow, OK. 😞